[docs] Explain an example read ambiguity scenario with a diagram. (#23256)
yugabyte/yugabyte-db: docs/content/preview/architecture/transactions/read-restart-error.md

YugabyteDB doesn't require atomic clocks, but instead allows a configurable setting for maximum clock skew. Time synchronization protocols such as NTP synchronize commodity hardware clocks periodically to keep the skew low and bounded. Additionally, there are optimizations in YugabyteDB to resolve this ambiguity internally with best-effort. However, when it can't be resolved internally, YugabyteDB will output a `read restart` error to the external client, similar to the following:

```output
ERROR:  Query error: Restart read required at: { read: { physical: 1656351408684482 } local_limit: { physical: 1656351408684482 } global_limit: <min> in_txn_limit: <max> serial_no: 0 }
```

A detailed scenario that explains how clock skew can result in the above mentioned ambiguity around data visibility is as follows:

* Consider a scenario where tokens 17, 29 are inserted into an empty tokens table. Then, all the tokens from the table are retrieved.

  SQL commands for the scenario:
  ```sql
  INSERT INTO tokens VALUES (17);
  INSERT INTO tokens VALUES (29);
  SELECT * FROM tokens;
  ```

> * The SELECT must return both 17, and 29. However, the clock skew can cause one or more tokens from the list to be omitted.

Author: Retaining `However, the clock skew can cause one or more tokens from the list to be omitted.`
I am open to rephrasing it but not removing it altogether.

* Due to clock skew, the INSERT operation can pick a commit time higher than the "true time". However, the SELECT picks a lower read time and thus omit the prior INSERT from the result set.

Follow the diagram below for a deeper discussion on the order of operations leading to this anomaly.

  ![Read Restart Error](/images/architecture/txn/read_restart_error.svg)

  In the above setup, there are three tablet servers, namely, `TSERVER 1`, `TSERVER 2`, and `TSERVER 3`. The data for the tokens table is hosted on servers TSERVER 2, and TSERVER 3. The query layer on TSERVER 1 is serving the SQL requests.

  Moreover, TSERVER 3's clock is running 5 units of time ahead of TSERVER 2's clock because of clock skew.

> 1. First, an INSERT of token 29 is issued to YSQL on TSERVER 1.

Reviewer: ```suggestion
1. An INSERT of token 29 is issued to YSQL on TSERVER 1.
```

> 2. This INSERT is routed to the tablet hosted on TSERVER 3. The operation picks a commit time of `T2=103` even though the "true time" is `98` (step 2).

Author: I could not make the arrows bigger. I zoomed into the picture instead. Let me know what you think.

3. TSERVER 1 acknowledges the INSERT.

> 4. Now, that the INSERT command is complete, the SELECT command is issued.

Reviewer: ```suggestion
4. Now that the INSERT command is complete, the SELECT command is issued.
```

8. Thus, token 29 is omitted from the result set returned.

  More generally, this anomaly occurs whenever the commit time of a prior write operation `T2` is higher than the read time `T1` of a later read operation, thus violating (distributed) consistency.

How does YugabyteDB mitigate this clock skew anomaly?

* First, note that the clock skew between all nodes in the cluster is always in a [max_clock_skew_usec](../../../reference/configuration/yb-tserver/#max-clock-skew-usec) bound due to clock synchronization algorithms.
* For records with a commit timestamp later than `T1` + `max_clock_skew`, the database can be sure that these records were written after the read was issued. But for records with a commit timestamp between `T1` and `T1` + `max_clock_skew`, the tablet servers can find themselves in an ambiguous situation, such as the following:

  * A read operation must observe all the commits that were acknowledged before the read was issued, even when the read time is picked on a node whose clock runs behind the node that committed the prior write. NOTE: the commit happened before the read as perceived by the user and not necessarily according to the machine clocks.

>   * A read operation must return a consistent snapshot of data, i.e. the read cannot simply advance its read time on observing a record with higher timestamp since the read already returned records from an older snapshot thus far. Therefore, the read must be restarted from the beginning with the advanced timestamp. Thus, the name read restart error.

Reviewer: ```suggestion
  * A read operation must return a consistent snapshot of data. That is, the read cannot simply advance its read time on observing a record with higher timestamp since the read already returned records from an older snapshot thus far. Therefore, the read must be restarted from the beginning with the advanced timestamp. Thus, the name read restart error.
```

>   * A read operation is free to include or exclude records committed by concurrent transactions. Implying, whenever the commit timestamp is lower than the read timestamp, the read is allowed to observe writes from the committed transaction, even when the commit happened concurrently with the read. Therefore, there is no ambiguity when the commit time is lower.

Reviewer: ```suggestion
  * A read operation is free to include or exclude records committed by concurrent transactions. This implies that whenever the commit timestamp is lower than the read timestamp, the read is allowed to observe writes from the committed transaction, even when the commit happened concurrently with the read. Therefore, there is no ambiguity when the commit time is lower.
```

* Whenever a read operation finds records with timestamp in the range `(T1, T1+max_clock_skew]`, to avoid breaking the strong guarantee that a reader should always be able to read what was committed earlier, and to read a consistent snapshot, the read operation raises a `Read restart` error to restart the read.

> {{<note title="Causal Reverse Anomaly">}}

Reviewer: What is a causal reverse anomaly? First mention. This seems out of place, why are you mentioning it here?
Or is this entire scenario a 'causal reverse anomaly', in which case why not call it that earlier?

Author: It is another anomaly that stems from clock skew. We do not handle this anomaly. This is a piece of info for those who are curious. But I do agree that there is no point in mentioning this since readers who know about this anomaly already know that YB does not handle it.

1. An INSERT of token 29 is issued to YSQL on TSERVER 1.
2. This INSERT is routed to the tablet hosted on TSERVER 3. The operation picks a commit time of `T2=103` even though the "true time" is `98`.
3. TSERVER 1 acknowledges the INSERT.
4. Now that the INSERT command is complete, the SELECT command is issued.
5. TSERVER 1 starts a distributed read. TSERVER 1 reads data from multiple shards on different physical YB-TServers in the cluster, namely TSERVER 2 and TSERVER 3. The read point that defines the snapshot of the database at which the data will be read, is picked on TSERVER 2 based on the safe time of that YB-TServer, namely `T1=101`.
6. TSERVER 1 gets back token 17 in the result set.
7. TSERVER 1 now issues a read operation to TSERVER 3 to retrieve records from the second shard. However, since the read time `T1=101` is less than the commit time of the prior insert `T2=103`, the record is not part of the read snapshot.
8. Thus, token 29 is omitted from the result set returned.

  More generally, this anomaly occurs whenever the commit time of a prior write operation `T2` is higher than the read time `T1` of a later read operation, thus violating (distributed) consistency.

> How does YugabyteDB mitigate this clock skew anomaly?

Reviewer: Would move this part up to after "YugabyteDB has optimizations to resolve this ambiguity internally with best-effort." given this appears to be an explanation of YBDB optimizations

Author: This is not one of the optimizations. We never describe the optimizations, just that they are present. Perhaps, mitigate is not the correct word to use here? I changed it to prevent. This basically describes how we raise a read restart error whenever we detect the anomaly. I hope that makes it clearer.
